coderd/workspacebuilds.go

```diff
@@ -389,6 +389,37 @@ func (api *API) postWorkspaceBuildsInternal(
 		builder = builder.Reason(database.BuildReason(createBuild.Reason))
 	}
 
+	// #20925: if the workspace is dormant and we are starting the workspace,
+	// we need to unset that status before inserting a new build.
+	if workspace.DormantAt.Valid && transition == database.WorkspaceTransitionStart {
+		if _, err := api.Database.UpdateWorkspaceDormantDeletingAt(ctx, database.UpdateWorkspaceDormantDeletingAtParams{
+			ID:        workspace.ID,
+			DormantAt: sql.NullTime{Valid: false},
+		}); err != nil {
+			return codersdk.WorkspaceBuild{}, httperror.NewResponseError(http.StatusInternalServerError, codersdk.Response{
+				Message: "Internal error unsetting workspace dormant status",
+				Detail:  err.Error(),
+			})
+		}
+		// We need to audit this change separately.
+		updatedWorkspace := workspace.WorkspaceTable()
+		updatedWorkspace.DormantAt = sql.NullTime{Valid: false}
+		auditor := api.Auditor.Load()
+		bag := audit.BaggageFromContext(ctx)
+		audit.BackgroundAudit(ctx, &audit.BackgroundAuditParams[database.WorkspaceTable]{
+			Audit:          *auditor,
+			Old:            workspace.WorkspaceTable(),
+			New:            updatedWorkspace,
+			Log:            api.Logger,
+			UserID:         apiKey.UserID,
+			OrganizationID: workspace.OrganizationID,
+			RequestID:      workspace.ID,
+			IP:             bag.IP,
+			Action:         database.AuditActionWrite,
+			Status:         http.StatusOK,
+		})
+	}
```

Comment:
Should this happen inside the transaction below? So if the workspace build cannot be created, the dormancy remains?

I can also see why the dormancy should be toggled regardless.